### PR TITLE
Max flow

### LINKED ISSUE #1712: Support the MaxFlow graph algorithm

### PR DESCRIPTION
**Pull request by swilly22**
Resolve: #1712

## Summary by CodeRabbit

* **New Features**
  * Added algo.maxFlow procedure to compute max flow with node/relationship filters, configurable capacity property, and selectable outputs (nodes, edges, edgeFlows, maxFlow). Now listed among available procedures.

* **Tests**
  * Added comprehensive test suite for algo.maxFlow covering validation, edge cases, multi-source/sink, capacities, filters, and output permutations.

* **Chores**
  * Updated an external submodule pointer.

* **Refactor**
  * Standardized integer types for array capacity/index/length APIs.

### CODE REVIEW

File: src/procedures/proc_maxflow.c [hunk 1]
```diff
@@ -0,0 +1,839 @@
...
+} MaxFlow_Context ;
+
+// context passed to the GraphBLAS IndexUnaryOp that maps each matrix entry to
+// its capacity value
+typedef struct {
+	const Graph *g;        // the graph being queried
+	AttributeID attr_id;   // attribute ID that holds the capacity value
+	SIType expected_type;  // accepted SIType(s) for the capacity attribute
+	uint64_t default_cap;  // capacity to use when the attribute is missing
+						   // or has the wrong type (default: 1)
```
Comment (GomezGab) [inline src/procedures/proc_maxflow.c]: This is a reasonable default when no attribute was passed in by the caller, but a confusing one if the caller asked for an attribute and there are edges with invalid attributes (in which case, I might expect 0 to be the default).

File: src/procedures/proc_maxflow.c [hunk 2]
```diff
@@ -0,0 +1,839 @@
...
+	GrB_OK (GrB_Matrix_nrows (&nrows, U)) ;
+	GrB_OK (GrB_Matrix_ncols (&ncols, U)) ;
+	GrB_OK (GrB_Matrix_new (&C, GrB_UINT64, nrows, ncols)) ;
+
+	GrB_OK (GrB_Matrix_apply_IndexOp_Scalar (
+		C, NULL, NULL, get_capacity, U, cap_ctx_s, NULL)) ;
+
+	GrB_OK (GrB_free (&cap_ctx_s)) ;
+	GrB_OK (GrB_free (&cap_ctx_type)) ;
+	GrB_OK (GrB_free (&get_capacity)) ;
```
Comment (GomezGab) [inline src/procedures/proc_maxflow.c:562]: `GrB_free(&U) ;`

File: src/procedures/proc_maxflow.c [hunk 3]
```diff
@@ -0,0 +1,839 @@
...
+	NodeID sink_id = INVALID_ENTITY_ID ;
+
+	bool multi_srcs  = array_len (srcs)  > 1 ;
+	bool multi_sinks = array_len (sinks) > 1 ;
+
+	if (!multi_srcs)  src_id  = ENTITY_GET_ID (srcs  [0]) ;
+	if (!multi_sinks) sink_id = ENTITY_GET_ID (sinks [0]) ;
+
+	if (multi_srcs || multi_sinks) {
+		// enlarge C to make room for a new source of sources and sink of sinks
```
Comment (GomezGab) [inline src/procedures/proc_maxflow.c:587]: Makes sense, but C is currently `RequiredMatrixDim` dimensions, so there may be a lot of extra space you could use. Does the extra space slow down `MaxFlow`? I'm not sure.

File: src/procedures/proc_maxflow.c [hunk 4]
```diff
@@ -0,0 +1,839 @@
...
+		sinks             == NULL ||
+		array_len (srcs)  == 0    ||
+		array_len (sinks) == 0
+	) {
+		ErrorCtx_SetError ("algo.maxFlow: expects at least "
+				"one source and one sink") ;
+
+		res = PROCEDURE_ERR ;
+		goto cleanup ;
+	}
```
Comment (GomezGab) [inline src/procedures/proc_maxflow.c:443]: srcs and sinks should be mutually exclusive. see bot's comment.

File: src/procedures/proc_maxflow.c [hunk 5]
```diff
@@ -0,0 +1,839 @@
...
+		GrB_OK (GrB_Matrix_ncols (&ncols, C)) ;
+		ASSERT (nrows == ncols) ;
+
+		GrB_Index n = nrows ;
+		if (multi_srcs)  n++ ;
+		if (multi_sinks) n++ ;
+
+		GrB_OK (GrB_Matrix_resize (C, n, n)) ;
+
+		uint64_t x = UINT32_MAX ;
```
Comment (GomezGab) [inline src/procedures/proc_maxflow.c]: why uint32 max and not uint64?

File: src/procedures/proc_maxflow.c [hunk 6]
```diff
@@ -0,0 +1,839 @@
...
+		}
+
+		if (multi_sinks) {
+			int l = array_len (sinks) ;
+			for (int i = 0 ; i < l ; i++) {
+				Node *s = sinks [i] ;
+				NodeID s_id = ENTITY_GET_ID (s) ;
+				GrB_OK (GrB_Matrix_removeElement (flow_mtx, s_id, sink_id)) ;
+			}
+		}
```
Comment (GomezGab) [inline src/procedures/proc_maxflow.c:699]: you could also 
```
		GrB_Index nrows, ncols ;
		GrB_OK (GrB_Matrix_nrows (&nrows, flow_mtx)) ;
		GrB_OK (GrB_Matrix_ncols (&ncols, flow_mtx)) ;
		ASSERT (nrows == ncols) ;

		GrB_Index n = nrows ;
		if (multi_srcs)  n-- ;
		if (multi_sinks) n-- ;

		GrB_OK (GrB_Matrix_resize (flow_mtx, n, n)) ;
```

File: src/procedures/proc_maxflow.c [hunk 7]
```diff
@@ -0,0 +1,839 @@
...
+		SIValue *nodes = array_new (SIValue, nvals) ;
+
+		// start collecting nodes
+		GxB_Iterator it ;
+		GrB_OK (GxB_Iterator_new (&it)) ;
+		GrB_OK (GxB_Vector_Iterator_attach (it, unique_nodes, NULL)) ;
+
+		GrB_Info info = GxB_Vector_Iterator_seek (it, 0) ;
+
+		while (info != GxB_EXHAUSTED) {
```
Comment (GomezGab) [inline src/procedures/proc_maxflow.c]: I prefer 
```
while (info == GrB_SUCCESS) {
 ...
}
ASSERT (info == GxB_EXHAUSTED)
```
So that you don't get stuck in a loop in case of failure

File: src/procedures/proc_maxflow.c [hunk 8]
```diff
@@ -0,0 +1,839 @@
...
+			// get the entry v(i)
+			GrB_Index node_id = GxB_Vector_Iterator_getIndex (it) ;
+
+			// fetch node
+			Node *n = rm_malloc (sizeof (Node)) ;
+			Graph_GetNode (g, node_id, n) ;
+			array_append (nodes, SI_Node (n)) ;
```
Comment (GomezGab) [inline src/procedures/proc_maxflow.c]: Kind of unfortunate to make a ton of tiny mallocs, also, I belive `SI_Node` will mark the allocation as `M_VOLATILE` and will not free. Does this leak? 

perhaps you should allocate an array of `Node *` of size `nvals`  and put it in the pdata so it gets freed after the procedure call?

File: src/procedures/proc_maxflow.c [hunk 9]
```diff
@@ -0,0 +1,839 @@
...
+				Edge *e = rm_malloc (sizeof (Edge)) ;
+				bool found = Graph_GetEdge (g, id, e) ;
+				ASSERT (found == true) ;
+
+				// initialize edge
+				e->src_id     = i ;
+				e->dest_id    = j ;
+				e->relationID = pdata->rel_id ;
+
+				array_append (edges, SI_Edge (e)) ;
```
Comment (GomezGab) [inline src/procedures/proc_maxflow.c]: same comment as with the nodes.

File: src/procedures/proc_maxflow.c (hunk 5 above)
Comment (GomezGab) [reply src/procedures/proc_maxflow.c]: Ideally this would be double infinity, since I think maxflow will cast to double anyway.

File: tests/flow/test_maxflow.py [hunk 10]
```diff
@@ -0,0 +1,791 @@
...
+                capacityProperty:  'cap',
+                relationshipTypes: ['PIPE']
+            })
+            YIELD maxFlow
+        """)
+        self.env.assertEqual(result.result_set[0][0], 9)
+
+    # ------------------------------------------------------------------ #
+    #  source == sink (degenerate case)                                  #
+    # ------------------------------------------------------------------ #
```
Comment (GomezGab) [inline tests/flow/test_maxflow.py:853]: Add a test for multi source / multisink where these are not disjoint.
ie. 
```
            CALL algo.maxFlow({
                sourceNodes:       [s, t],
                targetNodes:       [t],
                capacityProperty:  'cap',
                relationshipTypes: ['PIPE']
            })
```
```
            CALL algo.maxFlow({
                sourceNodes:       [s],
                targetNodes:       [s, t],
                capacityProperty:  'cap',
                relationshipTypes: ['PIPE']
            })
```
```
            CALL algo.maxFlow({
                sourceNodes:       [s0, s1, s2, x],
                targetNodes:       [t0, t1, t2, x],
                capacityProperty:  'cap',
                relationshipTypes: ['PIPE']
            })
```

File: tests/flow/test_maxflow.py [hunk 11]
```diff
@@ -0,0 +1,791 @@
...
+               CALL algo.maxFlow({sourceNodes:[s], targetNodes:[t],
+                   relationshipTypes:['R']})
+               YIELD maxFlow, badField""",
+
+            # extra / misspelled key alongside valid ones
+            """MATCH (s:N),(t:N) WHERE id(s)<>id(t)
+               CALL algo.maxFlow({sourceNodes:[s], targetNodes:[t],
+                   relationshipTypes:['R'], oops:'yes'})
+               YIELD maxFlow""",
+        ]
```
Comment (GomezGab) [inline tests/flow/test_maxflow.py:137]: Add a test to ensure matricies with tensors are rejected

**Review by GomezGab — CHANGES_REQUESTED** [review]
See sanitizer and bot comments aswell

File: src/procedures/proc_maxflow.c (hunk 2 above)
Comment (swilly22) [reply src/procedures/proc_maxflow.c:562]: Saved in the procedure context, do not free at this point in time.

File: src/procedures/proc_maxflow.c (hunk 1 above)
Comment (swilly22) [reply src/procedures/proc_maxflow.c]: Fail on negative values

Comment (swilly22) [reply src/procedures/proc_maxflow.c]: switch from integer to double, cast if needed.

File: src/procedures/proc_maxflow.c (hunk 6 above)
Comment (swilly22) [reply src/procedures/proc_maxflow.c:699]: resize, although cleaner, might be a bit more expensive due to memory allocations

File: src/procedures/proc_maxflow.c (hunk 7 above)
Comment (swilly22) [reply src/procedures/proc_maxflow.c]: I suggest following Tim's example from the GraphBLAS.h iter docs.

File: src/procedures/proc_maxflow.c (hunk 8 above)
Comment (swilly22) [reply src/procedures/proc_maxflow.c]: At the moment we don't have much of an option, later on we'll represent nodes/edges differently within an `SIValue`, also SI_Node and SI_Edge are setting allocation type to be volatile.

File: src/procedures/proc_maxflow.c (hunk 9 above)
Comment (swilly22) [reply src/procedures/proc_maxflow.c]: See comment above.

File: tests/flow/test_maxflow.py (hunk 11 above)
Comment (gkorland) [reply tests/flow/test_maxflow.py:137]: Added in 827fee0ae. New `test_01b_tensor_relationship_rejected` creates a relationship type with multi-edges (two R edges between the same pair of nodes) and verifies the procedure raises an error containing "multi-edges".

File: src/procedures/proc_maxflow.c (hunk 7 above)
Comment (gkorland) [reply src/procedures/proc_maxflow.c]: Fixed in a8be7ea6e. Both iterator loops (vector and matrix) now use `while (info == GrB_SUCCESS)` with `ASSERT(info == GxB_EXHAUSTED)` after the loop, following Tim's pattern from GraphBLAS.h docs. This prevents infinite loops on unexpected error codes in debug builds.

File: tests/flow/test_maxflow.py (hunk 10 above)
Comment (gkorland) [reply tests/flow/test_maxflow.py:853]: Added in 827fee0ae. Three new test cases cover the exact scenarios requested:
- `test_23b`: sourceNodes=[s,t], targetNodes=[t] — target also in sources
- `test_23c`: sourceNodes=[s], targetNodes=[s,t] — source also in targets
- `test_23d`: sourceNodes=[s0,s1,s2,x], targetNodes=[t0,t1,t2,x] — multi-set with shared node

All assert that the error contains "disjoint".

File: src/procedures/proc_maxflow.c (hunk 4 above)
Comment (gkorland) [reply src/procedures/proc_maxflow.c:443]: Implemented in the latest commits. Source and sink sets are now validated for disjointness (lines 443–482). For singleton sets a direct ID comparison is used; for larger sets a boolean vector is built for the bigger set and the smaller set is probed. Added tests: test_22 (single source==sink), test_23 (multi overlap), test_23b (source also in targets), test_23c (target also in sources), test_23d (multi-source/multi-sink with shared node).

File: src/procedures/proc_maxflow.c (hunk 6 above)
Comment (gkorland) [reply src/procedures/proc_maxflow.c:699]: Acknowledged. Keeping the element-by-element removal approach for now due to the performance concern with `GrB_Matrix_resize` (potential reallocation). The current approach is O(n_srcs + n_sinks) element removals which avoids any matrix memory reallocation.

File: src/procedures/proc_maxflow.c (hunk 9 above)
Comment (gkorland) [reply src/procedures/proc_maxflow.c]: Same as the node allocation comment above — edges use `rm_malloc` + `M_SELF` at line 842, so ownership is properly managed. The small-allocation overhead remains a known trade-off for future improvement.

File: src/procedures/proc_maxflow.c (hunk 8 above)
Comment (gkorland) [reply src/procedures/proc_maxflow.c]: Acknowledged. The current code uses `rm_malloc` + `SIValue_SetAllocationType(..., M_SELF)` at line 773, so each node allocation is owned by the SIValue and will be freed when the result set is released. The tiny-malloc overhead is noted — this can be improved in the future when nodes/edges get a lighter SIValue representation.

**Comment by gkorland** [comment]
## Agent Review Summary

### Comments Fixed (agreed & resolved)
- **Thread 21** (iterator loop pattern) → fixed in a8be7ea6e — both vector and matrix iterator loops now use `while (info == GrB_SUCCESS)` + `ASSERT(info == GxB_EXHAUSTED)` to prevent infinite loops on error
- **Thread 18** (source/sink disjointness) → already implemented; replied and resolved
- **Thread 24** (non-disjoint set tests) → added in 827fee0ae — three new tests: `test_23b`, `test_23c`, `test_23d` covering asymmetric source-in-targets, target-in-sources, and multi-set with shared node
- **Thread 25** (tensor rejection test) → added in 827fee0ae — `test_01b_tensor_relationship_rejected` verifies multi-edge relationships are rejected
- **Threads 22, 23** (node/edge allocation) → replied noting `M_SELF` ownership pattern; acknowledged as future improvement; resolved

### Comments Declined (disagreed — threads left open)
- **Thread 20** (use `GrB_Matrix_resize`) → replied acknowledging the trade-off; current element-by-element removal preferred for performance. Thread left open for reviewer.

### Outdated Threads (resolved)
- **Threads 15, 19** — code has been updated past these comments in prior commits; resolved as outdated.

### Tests Added
- `test_01b_tensor_relationship_rejected` — verifies tensor/multi-edge relationships are rejected
- `test_23b_source_also_in_targets` — source node appearing in targets → error
- `test_23c_target_also_in_sources` — target node appearing in sources → error
- `test_23d_multi_source_multi_sink_shared_node` — shared node in multi-sets → error

### CI Status
All CI failures are **pre-existing** (identical 17 failures on the prior commit d5d7b2e). No new failures introduced by these changes. Passing checks: spellcheck, CodeRabbit, Analyze, Check label or branch.

**Review by gkorland — APPROVED** [review]

**Review by GomezGab — APPROVED** [review]
Looks good

**Review by gkorland — COMMENTED** [review]
# Code Review

## Overview

Adds a new `algo.maxFlow` procedure (`src/procedures/proc_maxflow.{c,h}`, ~970 LOC) that runs LAGraph's max-flow on a label/relationship-filtered subgraph, with support for multi-source / multi-sink, custom capacity attributes, and a `defaultCapacity` fallback. Comes with a substantial test suite (`tests/flow/test_maxflow.py`, 36 tests). Minor cleanups in `array.{c,h}` (BSD-style `u_int32_t` → portable `uint32_t`), `procedure.c` registration, and one obsolete line removed from `test_betweenness.py`. The 250+ GraphBLAS PreJIT files are auto-generated noise — not the place to look for bugs.

## Correctness — Bugs

### 1. Memory leaks on error paths in `Proc_MaxFlowInvoke`
`proc_maxflow.c:564-571` and `proc_maxflow.c:643-651`:
```c
if (invalid_attributes) {
    ErrorCtx_SetError(...);
    res = PROCEDURE_ERR;
    rm_free(pdata);          // pdata->R = U was set on line 523
    goto cleanup;            // U (now orphaned) and C are both leaked
}
```
Same pattern in the capacity-range branch frees `C` but leaks `U` (held only in `pdata->R`). Free `pdata->R` (and `C` in the first case) before `rm_free(pdata)`.

### 2. Resource leak if `GrB_Matrix_apply_IndexOp_Scalar` aborts
`proc_maxflow.c:557-562`: the `cap_ctx_s`, `cap_ctx_type`, `get_capacity` frees are guarded by `GrB_OK` *after* the apply. If the apply fails, `GrB_OK` aborts the call and those three objects leak. Same pattern recurs throughout — the `GrB_OK` macro wrapping `GrB_free` is also defensively wrong, since `GrB_free` returns `GrB_SUCCESS` unconditionally and you don't want to short-circuit on it.

### 3. Misleading comment / dead-looking transpose
`proc_maxflow.c:502-523`: the comment claims "transpose into U so that U(i,j) represents the edge i→j", but `GrB_transpose(U, A, NULL, U, GrB_DESC_RST0)` uses `INP0=TRAN`, which negates the transpose — so this is effectively `U<S(A)> = U` with replace, i.e. a structural filter. The current behaviour is correct (U keeps `i→j` semantics so `src_id=i, dest_id=j` at line 864-865 works), but the comment is wrong and the operation is needlessly obscure. Prefer `GrB_Matrix_assign` with the mask, or document what is really happening.

### 4. `Build_Matrix` result `A` used as a structural mask only
`Build_Matrix(&A, ...)` populates `A` but its values are immediately discarded — only its structure feeds the mask. If `Build_Matrix` supports a structure-only mode, that would be cheaper.

### 5. `_get_edge_capacity` corner cases
`proc_maxflow.c:81-108`:
- `bool found = Graph_GetEdge(...); ASSERT(found == true);` — disabled in release builds; a failing edge lookup will then misbehave silently. Convert to a real error or confirm `ASSERT` is enabled in release here.
- `*z = (res == 1 && *z >= 0) ? *z : ctx->default_cap;` — when `SIValue_ToDouble` returns 0 (non-numeric), the value of `*z` is undefined yet read on the next line. The logic works because the branch picks `default_cap`, but the code is hard to follow.
- The `expected_type = T_INT64` field in `EdgeCapacityContext` is dead; remove it.

### 6. `arr_len` on `(SIValue*)args`
`proc_maxflow.c:391`: `arr_len((SIValue*)args)` — the `arr.h` macro reads metadata before the pointer; if the framework can ever pass a non-arr.h-allocated pointer this crashes. Verify `args` is always built with `arr_new`.

### 7. Empty-graph edge case for capacity-range check
`proc_maxflow.c:633-643`: after `GrB_Matrix_select_FP64`, if the matrix becomes empty the FP64 monoid reductions return their identity (`-INFINITY` for max, `+INFINITY` for min). Then `max >= min * UINT32_MAX` is `-inf >= inf`, false — so the check silently doesn't fire. Test 27 (zero-capacity edge) survives because at least one positive edge remains. Add an explicit `nvals` check or comment.

### 8. Race on `invalid_attributes`
`proc_maxflow.c:105-107`: reads `*(ctx->invalid)` with a plain dereference, only the store is atomic. The data race is benign (worst case is a redundant atomic_store), but use `atomic_load_explicit(..., memory_order_relaxed)` for cleanliness and to silence TSan.

## Correctness — Tests

- `test_03_max_flow_no_path_between_source_and_sink`: `if len(result.result_set) > 0:` makes the assertion conditional — masks a real bug. Behaviour should be deterministic; either always empty result or always `flow=0`.
- `test_02_max_flow_empty_graph` and the `try/except` blocks: catch the generic `Exception` and `if \"<text>\" not in str(e): raise`. Brittle to message wording and silently passes if it changes; consider matching exception type or asserting `expected in str(e)`.
- Test 19 docstring contradicts itself (\"flow = 8 ... -> flow = 3\") — clarify.

## Performance

- **Disjointness check** (`proc_maxflow.c:455-477`) allocates a `GrB_Vector` sized to `Graph_RequiredMatrixDim(g)` even for a handful of source/sink nodes. A `HashSet` keyed by `NodeID` would be O(n_big + n_small) without the dim-sized vector. Only matters for very large graphs with tiny source sets.
- **Magic constant `INT32_MAX` as super-edge capacity** (line 599): treats super-source/super-sink edges as effectively infinite. Combined with the `max >= min * 2^32` check, two distinct user capacities of e.g. `1` and `INT32_MAX/2` would be valid alone but adding the super-edges pushes the ratio over the limit. Run the range check *after* adding super-edges, or size the super-edge capacity to `sum(real capacities)` (the canonical safe choice).
- The capacity-range check uses `UINT32_MAX` (`2^32-1`) but the comment/error text says `2^32`. Tighten the wording or use `(double)UINT32_MAX + 1`.

## Style / Conventions

- `proc_maxflow.c:21` doc example: `defaultCapacity: ['CONNECTS']` should be a number, e.g. `defaultCapacity: 1`.
- Procedure-registration line `_procRegister(\"algo.maxFlow\", Proc_MaxFlowCtx) ;` uses trailing-space-`;` — inconsistent with neighbours.
- The `u_int32_t → uint32_t` rename in `array.{c,h}` is unrelated to maxflow; would be cleaner as a separate commit/PR.
- `double x =  INT32_MAX;` (line 599) has a double space; empty blank line at 600.
- `MAX_FLOW_CONFIG_KEY_COUNT = 6` magic — derive it from a key list or assert it on registration.

## Test Coverage

Coverage is broad: validation, empty graph, disconnected, simple/bottleneck/parallel paths, multi-source/sink combinations, label/rel filters, mixed numeric types, zero/small/large capacities, missing/wrong-type attributes with and without `defaultCapacity`, disjoint-set rejection, tensor rejection, capacity-range rejection. Good coverage for the feature. Add the deterministic empty-flow assertion once the conditional in `test_03` is fixed.

## Summary

Solid feature implementation with thorough tests. The blocking concerns are (1) the leaks on error paths in `Invoke` and (2) the misleading transpose comment / mechanics. Smaller nits around `INT32_MAX` super-edge capacity vs. range check, brittle test assertions, and the unrelated `uint32_t` rename. Once the leaks are patched and `test_03` is made deterministic, this is ready to merge.
